net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
+}
+
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.GetSingleLayer(ctx, cids, from)
+	if err != nil {
+		return nil, err
+	}
+	startingTipset, err := gsf.LoadTipsetFromCids(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	out := []types.TipSet{startingTipset}
+	isDone, err := done(startingTipset)
+	if err != nil {
+		return nil, err
+	}
+	if !isDone {
+		baseCid := cids[0]
+		recursionDepth := 1
+	makeRecursion:
+		for {
+			cidResponseLayers, err := gsf.GetBlocks(ctx, baseCid, from, recursionDepth)
+			if err != nil {
+				return nil, err
+			}
+			for _, cidResponseLayer := range cidResponseLayers {
+				ts, err := gsf.LoadTipsetFromCids(ctx, cidResponseLayer)
+				if err != nil {
+					return nil, err
+				}
+
+				out = append(out, ts)
+				isDone, err := done(ts)
+				if err != nil {
+					return nil, err
+				}
+
+				if isDone {
+					break makeRecursion
+				}
+				baseCid = cidResponseLayer[0]
+			}
+			if recursionDepth < maxRecursionDepth {
+				recursionDepth *= recursionMultiplier
+			}
+		}
+	}
+	return out, nil
+}
+
+// LoadTipsetFromCids loads and sanitizes an already downloaded TipSet
+// from an array of cids
+func (gsf *GraphSyncFetcher) LoadTipsetFromCids(ctx context.Context, cidResponseLayer []cid.Cid) (types.TipSet, error) {
+	blockResponseLayer := make([]blocks.Block, 0, len(cidResponseLayer))
+	for _, cid := range cidResponseLayer {
+		block, err := gsf.bs.Get(cid)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		blockResponseLayer = append(blockResponseLayer, block)
+	}
+	validatedBlocks, err := sanitizeBlocks(ctx, blockResponseLayer, gsf.validator)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	ts, err := types.NewTipSet(validatedBlocks...)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	return ts, nil
+}
+
+type resultSet struct {
+	responseChan <-chan graphsync.ResponseProgress
+	errChan      <-chan error
+}
+
+// GetSingleLayer requests a single layer of cids as individual bocks, fetching
+// non-recursively
+func (gsf *GraphSyncFetcher) GetSingleLayer(ctx context.Context, cids []cid.Cid, from peer.ID) error {
+	selector, err := gsf.bridge.BuildSelector(func(ssb selector.SelectorSpecBuilder) selector.SelectorSpec {
+		return ssb.Matcher()
+	})
+	if err != nil {
+		return err
+	}
+	resultSets := make([]resultSet, 0, len(cids))
+	for _, c := range cids {
+		responseChan, errChan := gsf.gs.Request(ctx, from, cidlink.Link{Cid: c}, selector)
+		resultSets = append(resultSets, resultSet{responseChan, errChan})
+	}
+	for _, rs := range resultSets {
+		for err := range rs.errChan {
+			return err
+		}
+		for range rs.responseChan {
+		}
```

Comment:
I think we don't need to consume these channels in full since their references don't escape this function, and they'll be collected, in which case this second loop is misleading. If we *do*, then this code is wrong because returning an error will cause leaks. Please add a comment either way.